Lib/fnmatch.py

```diff
@@ -86,10 +97,29 @@ def translate(pat):
         i = i+1
         if c == '*':
             # compress consecutive `*` into one
-            if (not res) or res[-1] is not STAR:
+            h = i-1
+            while i < n and pat[i] == '*':
+                i = i+1
+            if sep:
+                star_count = i-h
+                is_segment = (h == 0 or pat[h-1] == sep) and (i == n or pat[i] == sep)
+                if star_count == 1:
+                    if is_segment:
+                        add(f'{DOT}+')
+                    else:
+                        add(f'{DOT}*')
+                elif star_count == 2 and is_segment:
+                    if i == n:
+                        add('.*')
+                    else:
+                        add(f'(.*{SEP})?')
+                        i = i+1
+                else:
+                    raise ValueError("Invalid pattern: '**' can only be an entire path component")
```

Comment:
That code is really nice :o I'll look at adopting that approach I think!

Reply:
I'm glad you like it. In https://github.com/jaraco/zipp/pull/108, I've generalized the implementation to align more with some of the assumptions here. Supporting different separators added quite a bit more complexity because that one factor now ties deep into the implementation. I chose to implement a class to store that state, but nested functions or (less elegantly) parameter passing could also have been used. Addressing that one concern, much of the prior simplicity is lost.

Feel free to use that as inspiration, or if you'd like, I can work on contributing that as `glob.translate` (or similar). If we're heading that route, I'd also be willing to work on expanding the unit tests (as I requested of you for this implementation).

Reply:
I've been experimenting with your implementation and I think it's probably the way forward, but I had trouble getting all the pathlib tests passing. Perhaps we could look at adopting it after this PR lands?

Reply:
I've finally got this working using regular expressions. New PR: #109879